keps/sig-storage/4876-mutable-csinode-allocatable/README.md

```diff
@@ -0,0 +1,847 @@
+
+#### Story 2
+
+As a cluster operator, I use volumes during node setup for components like kubelet, containerd, and additional drivers. These boot volumes, which are not managed by CSI, may be detached after setup, and I need a way to reclaim these slots for other uses. The current static capacity reporting doesn't allow for this flexibility.
+
+#### Story 3
+
+As a cluster operator, I need the Kubernetes scheduler to accurately count the number of available device slots for both storage volumes and network interfaces. On certain machine types, network interfaces and volumes share device slots, and network interfaces may be dynamically attached after the CSI driver is registered. This results in an inaccurate `Allocatable.Count` for volumes, causing stateful pods to be scheduled on nodes with insufficient capacity, ultimately getting stuck in a `ContainerCreating` state.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+The following risks are identified:
+
+- Frequent updates/retrieval of the `CSINode` object could increase API server load.
+
+- Frequent calls to a CSI driver's `NodeGetInfo` RPC endpoint may become expensive, particularly if the operation involves retrieving information from a remote server or performing resource-intensive tasks. Specifically, this is a concern at scale, where the cumulative cost of multiple nodes repeatedly querying for updates is more impactful.
+
+- There's a race condition where the scheduler might assign a stateful pod to a node with insufficient capacity if the `CSINode.Spec.Drivers[*].Allocatable.Count` value hasn't been updated in time.
+
+The risks are mitigated as follows:
+
+- The use of the Kubernetes informer pattern in the scheduler. The scheduler uses a `CSINode` informer and lister to efficiently access and watch `CSINode` objects (this logic is already present).
+
+- Allow users to opt in to this feature at a per-CSI driver granularity by configuring the `CSIDriver` object. Specifically, administrators will be able to fine-tune the interval update value via the `NodeAllocatableUpdatePeriodSeconds` attribute in the `CSIDriver` object as per their specific requirement.
+
+- A reactive update mechanism is implemented to immediately update the `CSINode.Spec.Drivers[*].Allocatable.Count` value if a pod fails to enter a running state due volume attachment failures as a result of insufficient capacity. This ensures that even if a race occurs, Kubernetes quickly corrects itself and prevents further scheduling errors.
+
+## Design Details
```

Comment:
One more thing to check in the scheduler code. The scheduler will queue a pod for retry based on informer events. It may currently assume that CSINode is immutable. So that logic may need to be updated.

Reply:
I see, if we want the scheduler to re-queue pods reactively whenever `Allocatable.Count` changes, we need to also include:

```
{Event: framework.ClusterEvent{Resource: framework.CSINode, ActionType: framework.Update}}
```

in [EventsToRegister()](https://github.com/kubernetes/kubernetes/blob/8294abc599696e0d1b5aa734afa7ae1e4f5059a0/pkg/scheduler/framework/plugins/nodevolumelimits/csi.go#L85). 

Thank you for pointing this out! I'll include this detail in the KEP.